src/components/qas/questions/qaCalculator/index.jsx

```diff
@@ -1,66 +1,246 @@
+    const { id } = question;
+
+    this.checkAnswerable(answerable, activeId === id);
+  }
+
+  getAnswersFromData(data) {
+    const { question } = this.props;
+    const { questionType } = question || {};
+    const { value } = this.calculator[questionType];
+
+    return {
+      ...value,
+      ...data,
+    };
+  }
+
+  checkAnswerable(answerable, active) {
+    if (!answerable && active) {
+      this.setState(prevState => ({
+        ...prevState,
+        answerable: true,
+      }));
+    }
+  }
+
+  updateFocus(focusStatus) {
+    const { question, answerHandler } = this.props;
+    const { id } = question;
+    const { value } = this.state;
+
+    answerHandler(id, value || ' ', focusStatus ? 'focus' : 'blur');
+
     this.setState(prevState => ({
       ...prevState,
-      hasFocus,
+      hasFocus: focusStatus,
     }));
+  }
+
+  handleBlur = () => {
+    this.updateFocus(false);
   };
 
-  render() {
+  handleFocus = () => {
+    this.updateFocus(true);
+  };
+
+  getNewVal(value, valType) {
     const { question } = this.props;
-    const { hasFocus } = this.state;
+    const { questionType } = question || {};
+    const { value: oldValue } = this.state;
+    const { solutionVariable, formula } = this.calculator[questionType];
+
+    if (!value || !valType) return oldValue;
+
+    if (!solutionVariable) return formula(value);
+
+    const newVal = {
+      ...oldValue,
+      [valType]: value,
+    };
+
+    return {
+      ...newVal,
+      [solutionVariable]: formula(newVal),
+    };
+  }
+
+  handleChange = (value, valType) => {
+    const { question, answerHandler } = this.props;
+    const { id } = question;
+
+    this.setState(
+      prevState => ({
+        ...prevState,
+        hasFocus: true,
+        value: {
+          ...this.getNewVal(+value, valType),
+        },
+      }),
+      () => {
+        const { value: updatedVal } = this.state;
+        answerHandler(id, updatedVal, 'change');
+      }
+    );
+  };
+
+  render() {
+    const { question, answer, activeId } = this.props;
+    const { answerable, value, hasFocus } = this.state;
 
     const { questionType } = question;
+
+    const { id, label } = question;
+    const active = activeId === id;
+    const answered = !isEmpty(answer);
+    const answeredClasses = {
+      answered,
+      unanswered: !answered,
+      [answerableStyle]: answerable || answered || active,
+      answerable: answerable || answered || active,
+    };
     const cardClasses = classnames(qaCard, qaCalc, marginTop, {
-      [active]: hasFocus,
+      [activeClass]: hasFocus,
     });
+    const fieldClasses = classnames('qa-text-input', answeredClasses);
+    const labelClasses = classnames(calcLabel, answeredClasses);
+
+    const { inputs, equation } = this.calculator[questionType];
+    const FindEquation = equation || null;
 
-    const QACalc = this.calculators[questionType];
+    const isDistCalc = includes(questionType, 'DistanceCalculator');
```

Comment:
Not sure `includes` is necessary.  Should probably just be a normal comparison of strings if possible.  Additionally I'll point out that the need for this `isDistCalc` variable might bely one of the shortcomings of this approach.  For instance, what happens if there's another calculator with an expanding bit?

Reply:
I agree, as this was a thought of mine while I was updating this. I'm still not certain how I'd like to handle this going forward and thought that if the case ever came up I might have a better mind to solve this.